### PR TITLE
feat: load and display ZIM dictionary icon

### PR DESCRIPTION
**Pull request by shenlebantongying**
close https://github.com/xiaoyifang/goldendict-ng/issues/872

### CODE REVIEW

File: src/dict/zim.cc [hunk 1]
```diff
@@ -275,16 +277,30 @@ void ZimDictionary::loadIcon() noexcept
...
+  }
 
-  if( !loadIconFromFile( fileName ) )
-  {
-    // Load failed -- use default icons
-    dictionaryNativeIcon = dictionaryIcon = QIcon(":/icons/icon32_zim.png");
+  // Try to load zim's illustration, which is usually 48x48 png
+  try {
+    auto illustration = df.getIllustrationItem( 48 ).getData();
+    QImage img = QImage::fromData( reinterpret_cast< const uchar * >( illustration.data() ), illustration.size() );
```
Comment (xiaoyifang) [inline src/dict/zim.cc:292]: check illustration!=null?

File: src/dict/zim.cc [hunk 2]
```diff
@@ -275,16 +277,30 @@ void ZimDictionary::loadIcon() noexcept
...
+    dictionaryNativeIcon = dictionaryIcon = QIcon( QPixmap::fromImage( img ) );
```
Comment (xiaoyifang) [inline src/dict/zim.cc:293]: check img!=null

**Comment by xiaoyifang** [comment]
image

**Comment by shenlebantongying** [comment]
I think the warning is for embedded devices where memory is truly limited. When out of memory, the malloc and new will fail and causes problems too. If the user's computer runs out of memory, he may have much bigger problems. 

image
https://www.man7.org/linux/man-pages/man3/malloc.3.html

However, we can add some checks to be more robust.

**Comment by xiaoyifang** [comment]
QImage img  may `=null`

when img=null ,has to load the default icon as well.

shenleban tongying ***@***.***> 于2023年6月19日周一 10:21写道：

**Pull request merged**